packages/core/src/workflow/context.ts

```diff
@@ -1,20 +1,81 @@
-import { type EventTypes, type WorkflowEvent } from "./events";
-import { type StepFunction, type Workflow } from "./workflow";
+import { assertExists } from "../utils";
+import {
+  type EventTypes,
+  StartEvent,
+  StopEvent,
+  WorkflowEvent,
+} from "./events";
+
+export type StepFunction<
+  Start = string,
+  T extends WorkflowEvent = WorkflowEvent,
+> = (context: Context<Start>, ev: T) => Promise<WorkflowEvent>;
+
+export type StepMap = Map<
+  StepFunction<any>,
+  { inputs: EventTypes[]; outputs: EventTypes[] | undefined }
+>;
+
+export type ReadonlyStepMap = ReadonlyMap<
+  StepFunction<any>,
+  { inputs: EventTypes[]; outputs: EventTypes[] | undefined }
+>;
+
+export type ContextParams<Start> = {
+  startEvent: StartEvent<Start>;
+  steps: StepMap;
+  timeout: number | null;
+  verbose: boolean;
+};
+
+export class Context<Start = string>
+  implements AsyncIterable<WorkflowEvent, void, void>, Promise<StopEvent>
+{
+  readonly #steps: ReadonlyStepMap;
+  // reverse map of #steps, helper for get the next step
+  readonly #eventMap: WeakMap<typeof WorkflowEvent, StepFunction<Start>>;
+
+  readonly #startEvent: StartEvent<Start>;
+  readonly #queue: WorkflowEvent[] = [];
+  readonly #globals: Map<string, any> = new Map();
 
-export class Context {
-  #workflow: Workflow;
-  #queues: Map<StepFunction, WorkflowEvent[]> = new Map();
   #eventBuffer: Map<EventTypes, WorkflowEvent[]> = new Map();
-  #globals: Map<string, any> = new Map();
-  #streamingQueue: WorkflowEvent[] = [];
-  running: boolean = true;
+
+  #running: boolean = true;
+  #timeout: number | null = null;
   #verbose: boolean = false;
 
-  constructor(params: { workflow: Workflow; verbose?: boolean }) {
-    this.#workflow = params.workflow;
+  #getStepFunction(event: WorkflowEvent): StepFunction<Start> | undefined {
+    return this.#eventMap.get(event.constructor as EventTypes);
+  }
+
+  get running(): boolean {
+    return this.#running;
+  }
+
+  constructor(params: ContextParams<Start>) {
+    this.#steps = params.steps;
+    this.#startEvent = params.startEvent;
+    // init eventMap
+    this.#eventMap = new WeakMap();
+    for (const [step, { inputs }] of this.#steps) {
+      for (const input of inputs) {
+        this.#eventMap.set(input, step);
+      }
+    }
+
+    if (typeof params.timeout === "number") {
+      this.#timeout = params.timeout;
+    }
     this.#verbose = params.verbose ?? false;
+
+    // push start event to the queue
+    const step = this.#getStepFunction(this.#startEvent);
+    assertExists(step, "No step found for start event");
+    this.#pushEvent(this.#startEvent, step);
   }
 
+  // these two API is not type safe
```

Comment:
since get/set can can be used to store anything, I don't think it can be type safe?